temporal_sqlalchemy/bases.py

```diff
@@ -9,18 +9,71 @@
 import sqlalchemy as sa
 import sqlalchemy.dialects.postgresql as sap
 import sqlalchemy.orm as orm
+import sqlalchemy.orm.base as base
 import sqlalchemy.orm.attributes as attributes
 import psycopg2.extras as psql_extras
 
 from temporal_sqlalchemy import nine
-from temporal_sqlalchemy.metadata import get_session_metadata
 
-_ClockSet = collections.namedtuple('_ClockSet', ('effective', 'vclock'))
+_PersistentClockPair = collections.namedtuple('_PersistentClockPairs',
+                                              ('effective', 'vclock'))
 
 T_PROPS = typing.TypeVar(
     'T_PROP', orm.RelationshipProperty, orm.ColumnProperty)
 
 
+class ActivityState:
+    def __set__(self, instance, value):
+        assert instance.temporal_options.activity_cls, "Make this better Joey"
+        # TODO should not be able to change activity once changes have been made to temporal properties
+        setattr(instance, '__temporal_current_activity', value)
+
+        if value:
+            current_clock = instance.current_clock
+            current_clock.activity = value
+
+    def __get__(self, instance, owner):
+        if not instance:
+            return self
+
+        return getattr(instance, '__temporal_current_activity', None)
+
+    @staticmethod
+    def reset_activity(target, attr):
+        target.activity = None
+
+    @staticmethod
+    def activity_required(target, value, oldvalue, initiator):
+        if not target.activity and oldvalue is not base.NEVER_SET:
+            raise ValueError("activity required")
+
+
+class ClockState:
+    def __set__(self, instance, value: 'EntityClock'):
+        setattr(instance, '__temporal_current_tick', value)
+        if value:
+            instance.clock.append(value)
+
+    def __get__(self, instance, owner):
+        if not instance:
+            return self
+
+        vclock = getattr(instance, 'vclock') or 0
```

Comment:
Using `getattr` with a string constant like this is equivalent to `instance.vclock` and will still crash with an `AttributeError` unless you pass a default value as the third argument. Did you mean to do that?

```py
vclock = getattr(instance, 'vclock', 0)
```